### PR TITLE
Add support for SSE-KMS and DSSE-KMS for Mountpoint

### PR DESCRIPTION
**Pull request by sauraank**
## Description of change
Added KMS key type and their IDs as bucket option. Added these headers for Put Object. Still need to get these keys as PutObjectParams to be used by PutObject.
Also need to add tests and changelog.

Relevant issues: 
#534 
## Does this change impact existing behavior?
Yes, there will be added mount options when this change will get merged.

Yes. It will be a breaking change.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3-client/src/object_client.rs [hunk 1]
```diff
@@ -266,6 +266,13 @@ pub enum GetObjectAttributesError {
     NoSuchKey,
 }
 
+/// Type of Server side encryption(SSE) opted for Put Object
+#[derive(Debug, Clone)]
+pub enum KmsKeys {
+    SseKms,
+    DsseKms,
+}
```
Comment (dannycjones) [inline mountpoint-s3-client/src/object_client.rs:274]: This isn't KMS type but rather Server-side encryption (SSE) type.

We might include SSE-S3, SSE-C, or something new in the future.

File: mountpoint-s3/src/fs.rs [hunk 2]
```diff
@@ -307,6 +308,10 @@ pub struct S3FilesystemConfig {
     pub allow_delete: bool,
     /// Storage class to be used for new object uploads
     pub storage_class: Option<String>,
+    /// KMS key type for Server side encryption
+    pub kms_key: Option<KmsKeys>,
+    /// KMS key ID for Server side encryption
+    pub key_id: Option<String>
```
Comment (dannycjones) [inline mountpoint-s3/src/fs.rs:314]: I think we probably want an optional enum "Encryption"/"EncryptionConfig"/something better named which can have variants. Both variants will have an optional key ID.

```suggestion
    /// Server-side encryption for new S3 objects, if specified
    pub encryption: Option<Encryption>,
```

I see we also have something similar in `mountpoint-s3-client`... I'm less sure what we want to do there, maybe we want to specifically take each as a field like a normal SDK would do. **Let's discuss this as a team.**

File: mountpoint-s3/src/main.rs [hunk 3]
```diff
@@ -119,6 +120,31 @@ struct CliArgs {
     )]
     pub allow_other: bool,
 
+    #[clap(
+        long,
+        help = "Add the SSE KMS encryption for PUT requests",
+        help_heading = BUCKET_OPTIONS_HEADER,
+        group = "sse_options"
+    )]
+    pub sse_kms: bool,
```
Comment (dannycjones) [inline mountpoint-s3/src/main.rs:129]:
```suggestion
    #[clap(
        long,
        help = "Specify SSE-KMS encryption for new objects",
        help_heading = BUCKET_OPTIONS_HEADER,
        group = "sse_options"
    )]
    pub sse_kms: bool,
```

File: mountpoint-s3/src/main.rs [hunk 4]
```diff
@@ -119,6 +120,31 @@ struct CliArgs {
...
+
+    // Keeping [sse_kms] and [dsse_kms] in same arg group makes them mutually exclusive arguments
+    #[clap(
+        long,
+        help = "Add the DSSE KMS encryption for PUT requests",
+        help_heading = BUCKET_OPTIONS_HEADER,
+        group = "sse_options"
+    )]
+    pub dsse_kms: bool,
```
Comment (dannycjones) [inline mountpoint-s3/src/main.rs:138]: Drop the comment, or you can instead define the group above the struct like this:

```rust
#[clap(group(
    clap::ArgGroup::new("sse_options")
        .multiple(false)
        .args(&["dsse_kms", "sse_kms"]),
))]
```

Plus update description alike sse_kms.

**Review by dannycjones — COMMENTED** [review]
Nice start, provided some early feedback.

I think we want to review the CLI arguments themselves as well as the approach for the interface in `mountpoint-s3-client` as a team - let's discuss.

File: mountpoint-s3-client/src/lib.rs [hunk 5]
```diff
@@ -51,7 +51,7 @@ pub mod failure_client;
 pub mod imds_crt_client;
 #[doc(hidden)]
 pub mod mock_client;
-mod object_client;
+pub mod object_client;
```
Comment (passaro) [inline mountpoint-s3-client/src/lib.rs:54]: Revert and instead re-export the required type(s) explicitly under the `types` module (see from line 68).

File: mountpoint-s3-client/src/object_client.rs (hunk 1 above)
Comment (jamesbornholt) [reply mountpoint-s3-client/src/object_client.rs:274]: Each of these variants will potentially have a different set of config associated with it. So maybe make this a richer enum type:
```rust
pub enum ServerSideEncryption {
    /// Use the default configuration for the bucket
    #[default]
    Default,
    Kms {
        key_id: Option<String>,
    },
    DualLayerKms {
        key_id: Option<String>,
    },
    // not now, but in the future
    CustomerProvided {
        key: Key,
    }
}
```

**Comment by sauraank** [comment]
As @vladem is already has a PR #693  up for it, closing this PR.

**Pull request closed without merging**